src/binary_annotator.cpp

```diff
@@ -0,0 +1,768 @@
+          offset = BuildStruct(field_offset, regions, next_object);
+        } else {
+          const uint64_t next_object_offset =
+              field_offset + GetScalar<uint32_t>(field_offset);
+          regions.emplace_back(MakeBinaryRegion(
+              field_offset, sizeof(uint32_t), BinaryRegionType::UOffset, 0,
+              next_object_offset,
+              std::string("offset to field `") + field->name()->c_str() + "`"));
+          offset += sizeof(uint32_t);
+
+          BuildTable(next_object_offset, BinarySectionType::Table, next_object);
+        }
+      } break;
+
+      case reflection::BaseType::String: {
+        const uint64_t string_offset =
+            field_offset + GetScalar<uint32_t>(field_offset);
+
+        regions.emplace_back(MakeBinaryRegion(
+            field_offset, sizeof(uint32_t), BinaryRegionType::UOffset, 0,
+            string_offset,
+            std::string("offset to field `") + field->name()->c_str() + "`"));
+
+        BuildString(string_offset, table, field);
+      } break;
+
+      case reflection::BaseType::Vector: {
+        const uint64_t vector_offset =
+            field_offset + GetScalar<uint32_t>(field_offset);
+
+        regions.emplace_back(MakeBinaryRegion(
+            field_offset, sizeof(uint32_t), BinaryRegionType::UOffset, 0,
+            vector_offset,
+            std::string("offset to field `") + field->name()->c_str() + "`"));
+
+        BuildVector(vector_offset, table, field, table_offset, vtable);
+      } break;
+
+      case reflection::BaseType::Union: {
+        const uint64_t union_offset =
+            field_offset + GetScalar<uint32_t>(field_offset);
+
+        // The union type field is always one less than the union itself.
+        const uint16_t union_type_id = field->id() - 1;
+
+        auto vtable_entry = vtable.fields.find(union_type_id);
+        if (vtable_entry == vtable.fields.end()) {
+          // TODO(dbaileychess): need to capture this error condition.
+          break;
+        }
+
+        const uint64_t type_offset =
+            table_offset + vtable_entry->second.offset_from_table;
+
+        const uint8_t realized_type = GetScalar<uint8_t>(type_offset);
+
+        const std::string enum_type =
+            BuildUnion(union_offset, realized_type, field);
+
+        regions.emplace_back(MakeBinaryRegion(
+            field_offset, sizeof(uint32_t), BinaryRegionType::UOffset, 0,
+            union_offset,
+            std::string("offset to field `") + field->name()->c_str() +
+                "` (union of type `" + enum_type + "`)"));
+
+      } break;
+
+      default: break;
+    }
+  }
+
+  // Fill in any regions that weren't covered above, as those are padding
+  // regions.
+  size_t region_index = 1;
+  std::vector<BinaryRegion> padding_regions;
+  uint64_t i = field_offset_start;
+  while (region_index < regions.size() && i < table_end_offset) {
+    const uint64_t region_start = regions[region_index].offset;
```

Comment:
couldn't there be padding before the first region of the table?

Like if the minimum field of the vtable is > 4 then there's space between the table->vtable ptr and the first field's value.

Also, I hate to pile on the work but we probably should have unit tests for these kinds of cases

Reply:
This should be doing that? I start at region 1 since the first region is the vtable-offset and there should be no padding before that (or if there is, its going to be outputted as a new padding section in the final pass through the buffer). So with i = 1, I am looking at the first "field" of the table and the check below sees if there is any gap between the `field_offset_start` and that region's offset.

Reply:
Gotcha, makes sense. I mistook the first region to be the first field

Reply:
I'll add a comment to clarify that.